### PR TITLE
GetObject API with Args & Response objects.

### PR DESCRIPTION
**Pull request by BigUstad**
Versioning related code included. Examples added for versioning & conditional matches.

### CODE REVIEW

File: Minio.Examples/Cases/GetObjectQuery.cs [hunk 1]
```diff
@@ -0,0 +1,57 @@
+/*
+ * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2017 MinIO, Inc.
```
Comment (kannappanr) [inline Minio.Examples/Cases/GetObjectQuery.cs:2]:
```suggestion
 * MinIO .NET Library for Amazon S3 Compatible Cloud Storage, (C) 2020 MinIO, Inc.
```

File: Minio/DataModel/ServerSideEncryption.cs [hunk 2]
```diff
@@ -67,6 +69,11 @@ public SSEC(byte[] key)
             }
             this.key = key;
         }
+
+        public string GetCustomerKeyMD5()
+        {
+            return utils.getMD5SumStr(this.key);
+        }
```
Comment (poornas) [inline Minio/DataModel/ServerSideEncryption.cs:76]: Didn't we decide in previous reviews that this method is unnecessary

**Review by poornas — CHANGES_REQUESTED** [review]

File: Minio/DataModel/ServerSideEncryption.cs (hunk 2 above)
Comment (BigUstad) [reply Minio/DataModel/ServerSideEncryption.cs:76]: I've not pushed the resolved conflicts files yet. I will do it soon.

File: Minio.Functional.Tests/FunctionalTest.cs [hunk 3]
```diff
@@ -51,9 +51,9 @@ public class FunctionalTest
...
-        private const string getObjectSignature1 = "Task GetObjectAsync(string bucketName, string objectName, Action<Stream> callback, CancellationToken cancellationToken = default(CancellationToken))";
-        private const string getObjectSignature2 = "Task GetObjectAsync(string bucketName, string objectName, Action<Stream> callback, CancellationToken cancellationToken = default(CancellationToken))";
-        private const string getObjectSignature3 = "Task GetObjectAsync(string bucketName, string objectName, string fileName, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string getObjectSignature1 = "Task GetObjectAsync(GetObjectArgs args, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string getObjectSignature2 = "Task GetObjectAsync(GetObjectArgs args, CancellationToken cancellationToken = default(CancellationToken))";
+        private const string getObjectSignature3 = "Task GetObjectAsync(GetObjectArgs args, CancellationToken cancellationToken = default(CancellationToken))";
```
[the remark this answers is not in the document]
Comment (BigUstad) [reply Minio.Functional.Tests/FunctionalTest.cs:56]: I've not pushed the resolved conflicts files yet. I will do it soon.

**Comment by poornas** [comment]
can you rebase this PR @BigUstad

**Comment by BigUstad** [comment]
> can you rebase this PR @BigUstad

Yes @poornas I'm rebasing all PRs that needs merging.

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 4]
```diff
@@ -165,4 +167,86 @@ public override void Validate()
...
+            this.ObjectLength = length;
+            if (ObjectLength > 0)
+            {
+                this.HeaderMap.Add("Range", "bytes=" + offset.ToString() + "-" + (offset + length - 1).ToString());
+            }
+
+            if (this.SSE != null && this.SSE.GetType().Equals(EncryptionType.SSE_C))
+            {
+                this.SSE.Marshal(this.HeaderMap);
+            }
```
Comment (poornas) [inline Minio/DataModel/ObjectOperationsArgs.cs:242]: you are already doing this in BuildRequest, this if block is not needed

File: Minio/DataModel/ServerSideEncryption.cs (hunk 2 above)
Comment (poornas) [reply Minio/DataModel/ServerSideEncryption.cs:76]: not fixed

File: Minio/DataModel/ServerSideEncryption.cs [hunk 5]
```diff
@@ -102,6 +109,11 @@ public virtual void Marshal(Dictionary<string, string> headers)
             headers.Add(Constants.SSEGenericHeader, "AES256");
             return;
         }
+
+        public string GetCustomerKeyMD5()
+        {
+            return null;
+        }
```
Comment (poornas) [inline Minio/DataModel/ServerSideEncryption.cs:116]: this method is not required

File: Minio/DataModel/ServerSideEncryption.cs [hunk 6]
```diff
@@ -136,6 +148,11 @@ public void Marshal(Dictionary<string, string> headers)
             return;
         }
 
+        public string GetCustomerKeyMD5()
+        {
+            return null;
+        }
```
Comment (poornas) [inline Minio/DataModel/ServerSideEncryption.cs:154]: this method is not required.

File: Minio/Helper/OperationsHelper.cs [hunk 7]
```diff
@@ -15,11 +15,107 @@
...
 {
+    public partial class MinioClient : IObjectOperations
+    {
+        /// <summary>
+        /// private helper method to remove list of objects from bucket
+        /// </summary>
+        /// <param name="args">GetObjectArgs Arguments Object encapsulates information like - bucket name, object name etc <param>
+        /// <param name="objectStat"> ObjectStat object encapsulates information like - object name, size, etag etc <param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
+        private async Task getObjectFileAsync(GetObjectArgs args, ObjectStat objectStat, CancellationToken cancellationToken = default(CancellationToken))
```
Comment (poornas) [inline Minio/Helper/OperationsHelper.cs:37]: please  fix github actions reported issues

File: Minio/Helper/OperationsHelper.cs [hunk 8]
```diff
@@ -15,11 +15,107 @@
...
+            await getObjectStreamAsync(args, objectStat, null, cancellationToken).ConfigureAwait(false);
+        }
+
+
+        /// <summary>
+        /// private helper method to remove list of objects from bucket
+        /// </summary>
+        /// <param name="args">GetObjectArgs Arguments Object encapsulates information like - bucket name, object name etc <param>
+        /// <param name="objectStat"> ObjectStat object encapsulates information like - object name, size, etag etc <param>
+        /// <param name="cancellationToken">Optional cancellation token to cancel the operation</param>
```
Comment (poornas) [inline Minio/Helper/OperationsHelper.cs:111]: fix github actions

File: Minio.Functional.Tests/FunctionalTest.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (poornas) [reply Minio.Functional.Tests/FunctionalTest.cs:57]: not fixed

**Review by poornas — CHANGES_REQUESTED** [review]
github actions reported warnings/errors need to be fixed.

File: Minio/Helper/OperationsHelper.cs (hunk 7 above)
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:37]: These are now fixed.

File: Minio/Helper/OperationsHelper.cs (hunk 8 above)
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:112]: These are now fixed.

**Review by poornas — APPROVED** [review]
LGTM

File: Minio/Helper/OperationsHelper.cs [hunk 9]
```diff
@@ -15,11 +15,108 @@
...
+            {
+                tempFileName = $"{args.FileName}.{etag}.{args.VersionId}.part.minio";
+            }
+
+            bool tempFileExists = File.Exists(tempFileName);
+            bool getObjectFileExists = File.Exists(args.FileName);
+
+            utils.ValidateFile(tempFileName);
+
+            FileInfo tempFileInfo = new FileInfo(tempFileName);
```
Comment (kannappanr) [inline Minio/Helper/OperationsHelper.cs:53]: can this be moved inside the if condition code

Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:53]: Yes. Changed.

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 10]
```diff
@@ -363,7 +365,7 @@ public PresignedPutObjectArgs WithExpiry(int ex)
         }
     }
 
-    public class RemoveUploadArgs : EncryptionArgs<RemoveUploadArgs>
+public class RemoveUploadArgs : EncryptionArgs<RemoveUploadArgs>
```
Comment (kannappanr) [inline Minio/DataModel/ObjectOperationsArgs.cs:368]: indentation

Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Noted, changed.

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 11]
```diff
@@ -400,4 +402,81 @@ public RemoveIncompleteUploadArgs()
...
+            if (OffsetLengthSet)
+            {
+                if (this.ObjectOffset < 0)
+                {
+                    throw new ArgumentException("Offset should be zero or greater", nameof(this.ObjectOffset));
+                }
+
+                if (this.ObjectLength < 0)
+                {
+                    throw new ArgumentException("Length should be greater than zero", nameof(this.ObjectLength));
```
Comment (kannappanr) [inline Minio/DataModel/ObjectOperationsArgs.cs:436]: message should be greater than or equal to

File: Minio/DataModel/ObjectOperationsArgs.cs [hunk 12]
```diff
@@ -335,7 +337,7 @@ public PresignedPostPolicyArgs WithPolicy(PostPolicy policy)
         {
             this.Policy = policy;
             return this;
-        }
+        }
```
Comment (kannappanr) [inline Minio/DataModel/ObjectOperationsArgs.cs:340]: why is there an extra space

File: Minio/Helper/OperationsHelper.cs [hunk 13]
```diff
@@ -15,11 +15,107 @@
...
+                // We download for all other conditions
+            }
+
+            utils.ValidateFile(tempFileName);
+
+            if (tempFileExists)
+            {
+                FileInfo tempFileInfo = new FileInfo(tempFileName);
+                tempFileSize = tempFileInfo.Length;
+                if (tempFileSize > length)
```
Comment (kannappanr) [inline Minio/Helper/OperationsHelper.cs:78]: what if tempFileSize is < length?

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 12 above)
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs:340]: Happened during vscode 'Accept Both changes'. Removing.

File: Minio/DataModel/ObjectOperationsArgs.cs (hunk 11 above)
Comment (BigUstad) [reply Minio/DataModel/ObjectOperationsArgs.cs]: Changed the message & the condition.

File: Minio/Helper/OperationsHelper.cs [hunk 14]
```diff
@@ -15,11 +15,79 @@
...
+
+            long tempFileSize = 0;
+            string tempFileName = $"{args.FileName}.{etag}.part.minio";
+            if (!string.IsNullOrEmpty(args.VersionId))
+            {
+                tempFileName = $"{args.FileName}.{etag}.{args.VersionId}.part.minio";
+            }
+            if (File.Exists(args.FileName))
+            {
+                File.Delete(tempFileName);
```
Comment (kannappanr) [inline Minio/Helper/OperationsHelper.cs:50]:
```suggestion
                File.Delete(args.FileName);
```

File: Minio/Helper/OperationsHelper.cs [hunk 15]
```diff
@@ -15,11 +15,79 @@
...
+            }
+
+            utils.ValidateFile(tempFileName);
```
Comment (kannappanr) [inline Minio/Helper/OperationsHelper.cs:53]: what does this do?

File: Minio/Helper/OperationsHelper.cs (hunk 14 above)
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:50]: Changing that.

File: Minio/Helper/OperationsHelper.cs (hunk 15 above)
Comment (BigUstad) [reply Minio/Helper/OperationsHelper.cs:53]: Validation that was previously there. It is to validate name before proceeding to save, so as to not encounter error during File operation.

**Review by kannappanr — APPROVED** [review]
LGTM

**Pull request merged**